### PR TITLE
3.0 real unbuffered

### PR DESCRIPTION
**Pull request by lorenzo**
Implemented real non-buffered Database statements.
The old implementation was secretly relying on PHP results buffering for MySQL and it was impossible to turn off.

Also improved the Sqlite driver to only buffer results optionally as mysql does.

I haven't tried yet with Postgres and will see how it fails in SQL Server so I can see if it is implementable at all. Treat this as a WIP

### CODE REVIEW

File: src/Database/Query.php [hunk 1]
```diff
@@ -1582,6 +1590,33 @@ public function valueBinder($binder = null)
...
+     * buffered. This enables you to iterate a result set multiple times, or
+     * both cache and iterate it.
+     *
+     * When disabled it will consume less memory as fetched results are not
+     * remembered for future iterations.
+     *
+     * If called with no arguments, it will return whether or not buffering is
+     * enabled.
+     *
+     * @param bool $enable whether or not to enable buffering
```
Comment (dereuromark) [inline src/Database/Query.php]: bool|null

File: src/Database/Statement/SqliteStatement.php [hunk 2]
```diff
@@ -19,10 +19,35 @@
...
+        if ($this->_statement instanceof BufferedStatement) {
+            $this->_statement = $this->_statement->getInnerStatement();
+        }
+
+        if ($this->_bufferResults) {
+            $this->_statement = new BufferedStatement($this->_statement, $this->_driver);
+        }
+
+        $result = $this->_statement->execute($params);
+        return $result;
```
Comment (markstory) [inline src/Database/Statement/SqliteStatement.php]: This line and the line above could be combined.

**Comment by markstory** [comment]
Looks realy good so far.

**Comment by lorenzo** [comment]
This can be merged now, although I did not manage to get unbuffered results from postgres or sqlserver. For postgres it is just not possible without a lot of manual fork. I will give it another try at a later point for sqlserver

File: tests/TestCase/Database/QueryTest.php [hunk 3]
```diff
@@ -3132,6 +3132,40 @@ public function testSqlCaseStatement()
...
+     */
+    public function testUnbufferedQuery()
+    {
+        $query = new Query($this->connection);
+        $result = $query->select(['body', 'author_id'])
+            ->from('articles')
+            ->bufferResults(false)
+            ->execute();
+
+        $this->skipIf(
```
Comment (josegonzalez) [inline tests/TestCase/Database/QueryTest.php]: Shouldn't you skip at the beginning of the test?

Comment (lorenzo) [inline tests/TestCase/Database/QueryTest.php]: This is the laziest way for me. I don't need to test for specific drivers, only statement features

**Comment by markstory** [comment]
Would it be good to document the limitations for postgres/sqlserver in the book?

**Comment by lorenzo** [comment]
Yes @markstory we need to document those

**Comment by markstory** [comment]
I can take care of the docs as I have a few other docs changes to do tonight.

**Pull request merged**